### ISSUE TITLE
I2C endless SCL strobe until timeout on esp32-s3 wroom and esp-idf 4.4 (IDFGH-6923)

### ISSUE THREAD

**Issue by bar-bog** [espressif/esp-idf#8543]
## Environment
- Module or chip used:  ESP32-S3-WROOM-1
- IDF version: release/v4.4
- Build System: idf.py
- Compiler version : xtensa-esp32-elf-gcc (crosstool-NG esp-2021r2-patch3) 8.4.0
- Operating System:  Linux
- Using an IDE?: No
- Power Supply:  Bench power supply with 5A limit

## Problem Description

I have encountered an I2C issues when switching custom boards that had esp32 wroom modules on it with new esp32-s3 wroom. Issue happens only on esp32-s3 - when switching target to esp32 and running original board (with esp32) with the same code base the issue is not existing. 
Problem occurs when communicating with a I2C chip (SLG46826) and only when sending a command with one byte of data. Before this error there is correct communication with the same chip.
There was a few issues mentioned that are similar, but couldn't find a solution there and it's not exactly the same problem in my opinion:
https://github.com/espressif/esp-idf/issues/8510

### Expected Behavior
ESP32-S3 should stop clocking SCL line after STOP condition and return ESP_OK error.

### Actual Behavior
ESP32-S3 is clocking SCL line after receiving a STOP condition until a Timeout occurs after one second. I2C driver returns TIMEOUT error to the application layer.

### Logic analyzer images
Full communication with long clock pulse at the end:
![full_comm](https://user-images.githubusercontent.com/18262496/157631676-80f83328-adbf-4c4c-88cc-9e5b14c851be.png)
Correct communication before an error:
![correct_comm](https://user-images.githubusercontent.com/18262496/157631696-d4a761f2-8ca3-442f-9527-47c981ded394.png)
Good communication in green and incorrect one in red:
![good_bad](https://user-images.githubusercontent.com/18262496/157631659-350fc2f6-075d-413a-b602-b80f9b8f572f.png)

### Code to reproduce this issue
I am using this code as a function to write to sensor
```C
    esp_err_t err = ESP_FAIL;
    uint8_t addr;

    if (mem_reg == REGISTER_DATA_BLOCK) {
        addr = SLG_I2C_BASE_ADDRESS;
    } else if (mem_reg == NVM_DATA_BLOCK) {
        addr = SLG_I2C_BASE_ADDRESS | NVM_DATA_BLOCK_I2C;
    } else if (mem_reg == EEPROM_DATA_BLOCK) {
        addr = SLG_I2C_BASE_ADDRESS | EEPROM_DATA_BLOCK_I2C;
    } else {
        return err;
    }

    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
    i2c_master_start(cmd);
    i2c_master_write_byte(cmd, addr | WRITE_BIT, ACK_CHECK_EN);
    i2c_master_write_byte(cmd, reg, ACK_CHECK_EN);
    i2c_master_write(cmd, bufp, len, ACK_CHECK_EN);
    i2c_master_stop(cmd);
    err = i2c_master_cmd_begin(i2c_num, cmd, 1000 / portTICK_RATE_MS);
    i2c_cmd_link_delete(cmd);

    return err;
```

**Comment by bar-bog** [comment]
I think I was able to narrow down this issue to no interrupt after stop condition. The i2c state machine expects this interrupt but in this case when it's not being sent the next event that comes to the state machine is a timeout event. I have made a short fix for this that once there is no next i2c command to execute and the op_code of it is I2C_LL_CMD_STOP I'm sending a done event and let the stop command be executed:
```diff
         p_i2c->cmd_idx++;
         p_i2c->cmd_link.head = p_i2c->cmd_link.head->next;
-        if (p_i2c->cmd_link.head == NULL || p_i2c->cmd_idx >= 15) {
+        if (p_i2c->cmd_link.head == NULL && p_i2c->cmd_idx >= 15) {
             p_i2c->cmd_idx = 0;
             break;
         }
+        else if(p_i2c->cmd_link.head == NULL && hw_cmd.op_code == I2C_LL_CMD_STOP)
+        {
+			evt.type = I2C_CMD_EVT_DONE;
+			xQueueOverwriteFromISR(p_i2c->cmd_evt_queue, &evt, &HPTaskAwoken);
+			// Return to the IDLE status after cmd_eve_done signal were send out.
+			p_i2c->status = I2C_STATUS_IDLE;
+        }
```

Not really sure if that's a correct fix, but maybe it would be helpful.

**Comment by mythbuster5** [comment]
[i2c_patch_1210.txt](https://github.com/espressif/esp-idf/files/8261372/i2c_patch_1210.txt)

Could you please try with this patch? To see whether this can help, thanks!

**Comment by bar-bog** [comment]
I have checked out the fresh esp-idf v4.4 and tried to apply the patch but it seems like it does not apply:
```
esp-idf git:(8153bfe412) ✗ git apply i2c_patch_1210.txt 
error: patch failed: components/driver/i2c.c:531
error: components/driver/i2c.c: patch does not apply
error: patch failed: components/hal/esp32c3/include/hal/i2c_ll.h:799
error: components/hal/esp32c3/include/hal/i2c_ll.h: patch does not apply
error: patch failed: components/hal/i2c_hal_iram.c:16
error: components/hal/i2c_hal_iram.c: patch does not apply
```
Did you made the patch on 4.4 and for esp32-s3 ? I can still try to apply it manually if that would help you.

**Comment by mythbuster5** [comment]
> I have checked out the fresh esp-idf v4.4 and tried to apply the patch but it seems like it does not apply:
> 
> ```
> esp-idf git:(8153bfe412) ✗ git apply i2c_patch_1210.txt 
> error: patch failed: components/driver/i2c.c:531
> error: components/driver/i2c.c: patch does not apply
> error: patch failed: components/hal/esp32c3/include/hal/i2c_ll.h:799
> error: components/hal/esp32c3/include/hal/i2c_ll.h: patch does not apply
> error: patch failed: components/hal/i2c_hal_iram.c:16
> error: components/hal/i2c_hal_iram.c: patch does not apply
> ```
> 
> Did you made the patch on 4.4 and for esp32-s3 ? I can still try to apply it manually if that would help you.

Sorry, let me take a look

**Comment by mythbuster5** [comment]
[i2c_patch_0317.txt](https://github.com/espressif/esp-idf/files/8281369/i2c_patch_0317.txt)

Sorry, there are too many new changes, please try this one? Thanks !

**Comment by bar-bog** [comment]
Hi, sorry for delay and thank you for your help . I did try to apply your patch on both v4.4 tag and release/v4.4 but this patch still didn't apply for me. Tried to apply it manually but it didn't work for me too. I think I got your intent with the change correctly i.e. guard the i2c isr with interrupt mask and force non-hardware bus reset. However I'm still stuck at the transaction.
What I have seen in my test is that I'm having one interrupt to little (no interrupt at last operation from the list) so I'm doing it manually.

**Comment by AxelLin** [comment]
> [i2c_patch_0317.txt](https://github.com/espressif/esp-idf/files/8281369/i2c_patch_0317.txt)
> 
> Sorry, there are too many new changes, please try this one? Thanks !

@mythbuster5 
Above patch is against master, however this bug report is against v4.4. (It does not apply to v4.4 branch).

**Comment by AxelLin** [comment]
@bar-bog @mythbuster5 
How about let's sync to v4.4.1?
@mythbuster5 Then if you have patch for test, you can make the patch against v4.4.1.

**Comment by mythbuster5** [comment]
@AxelLin  Emm, I'm trying to figure out the root cause of this problem, so I want to know the information of the latest commit the user used.

Might not just a I2C issue, also probably interrupt handler issue~

**Comment by aswinda29** [comment]
Hi, I was having exactly the same issue with v4.4. Then I upgraded to the master and the issue continued occur. 

What worked for me was applying the above mentioned patch [i2c_patch_0317.txt](https://github.com/espressif/esp-idf/files/8281369/i2c_patch_0317.txt) coupled with the new i2c wrappers.

i2c_master_write_to_device()
i2c_master_read_from_device()
i2c_master_write_read_device()

Here we can define the exact size of the read and write buffer length. This fixed the error I was getting. I hope this might help someone.

**Comment by Alvin1Zhang** [comment]
Thanks for reporting, fix on release/4.4 is available at https://github.com/espressif/esp-idf/commit/f86be6141b3b92f84d0645cba94636ba8944c9f3, feel free to reopen if the issue still happens. Thanks.
